File: client/src/main/scala/ch/fhnw/ima/bimgur/service/RepositoryService.scala
package ch.fhnw.ima.bimgur.service

import cats.data.Xor
import ch.fhnw.ima.bimgur._
import ch.fhnw.ima.bimgur.model.activiti.{ProcessDefinition, ProcessDefinitionKey, ProcessDefinitionList}
import io.circe.generic.auto._
import io.circe.Error
import io.circe.parser._
import org.scalajs.dom.ext.Ajax

import scala.concurrent.Future
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue

case class RepositoryService(activitiRestUrl: String) {

  val url = activitiRestUrl + "/repository"

  def getProcessDefinitions(key: ProcessDefinitionKey): Future[Xor[Error, Seq[ProcessDefinition]]] = {
    val requestFuture = Ajax.get(s"$url/process-definitions?key=$key")
    requestFuture.map(response => decode[ProcessDefinitionList](response.responseText).map(_.data))
  }

}

File: bimgur-activiti-rest-client/src/main/java/ch/fhnw/ima/bimgur/activiti/service/RepositoryService.java
package ch.fhnw.ima.bimgur.activiti.service;

import ch.fhnw.ima.bimgur.activiti.model.ProcessDefinition;
import ch.fhnw.ima.bimgur.activiti.model.Deployment;
import ch.fhnw.ima.bimgur.activiti.model.util.ResultList;
import ch.fhnw.ima.bimgur.activiti.service.util.ResultListExtractor;
import io.reactivex.Observable;
import retrofit2.http.GET;

public interface RepositoryService {

    @GET("repository/process-definitions")
    Observable<ResultList<ProcessDefinition>> getProcessDefinitionList();

    default Observable<ProcessDefinition> getProcessDefinition() {
        return ResultListExtractor.extract(getProcessDefinitionList());
    }

    @GET("repository/deployments")
    Observable<ResultList<Deployment>> getDeploymentList();

    default Observable<Deployment> getDeployments() {
        return ResultListExtractor.extract(getDeploymentList());
    }


}